File: code/arrange_tmx_files_with_extension.js
const fs = require('fs');
const path = require('path');
const glob = require('glob');
const { JSDOM } = require('jsdom');
const axios = require('axios');
const fetch = require('sync-fetch');

// Program Description
const appDesc = "Arrange TMs by batch or domain on batch transition";

// Parse command line arguments
const args = process.argv.slice(2);
let rootDirPath = '';
let version = false;
args.forEach(arg => {
    if (arg === '-V' || arg === '--version') {
        version = true;
    } else if (arg === '-r' || arg === '--repo') {
        rootDirPath = args[args.indexOf(arg) + 1];
    }
});

// Show version if requested
const versionText = `${appDesc} v. 0.2`;
if (version) {
    console.log(`versionText=${versionText}`);
    process.exit();
}

// Check if root directory path is provided
if (!rootDirPath) {
    console.log("Required argument not found. Run this script with parameter `-r /path/to/repo`.");
    process.exit();
}

// Constants
const disallowedDomains = ['CRT', 'FLQ', 'FNL', 'WBQ'];
const allowedDomains = {
    QQS: ['STQ', 'STQ-UH', 'STQ-UO', 'ICQ'],
    QQA: ['SCQ', 'TCQ', 'PAQ'],
    COS: ['MAT', 'REA', 'SCI'],
};
const trendTag = "MS2022";
const newTag = "FT2025";

const tmDirPath = path.join(rootDirPath, 'tm')

const idleExtension = '.idle';
const locales = getLocales();
const batches = ["01_COS_SCI-A_N", "02_COS_SCI-B_N", "03_COS_SCI-C_N", "04_QQS_N", "05_QQA_N", "06_COS_LDW_N", "07_COS_XYZ_N", "08_CGA_SCI_N", "11_COS_MAT-A_T", "12_COS_MAT-B_T", "13_COS_REA-A_T", "14_COS_REA-B_T", "15_COS_SCI-A_T", "16_COS_SCI-B_T", "17_CGA_SCI_T", "18_CGA_MAT_T", "19_CGA_REA_T", "21_COSP_REA-A_T", "22_COSP_REA-B_T", "23_COSP_MAT-A_T", "24_COSP_MAT-B_T", "25_COSP_SCI-A_N", "26_COSP_SCI-A_T", "27_QQSP_N", "28_QQAP_N", "29_COSP_XYZ_N", "30_CGAP_MAT_T", "31_CGAP_REA_T", "32_CGAP_SCI_T", "33_CGAP_SCI_N"]

// Helper functions
function removeSuffix(str, suffix) {
    if (str.endsWith(suffix)) {
        return str.slice(0, -suffix.length);
    }
    return str;
}

function removePrefix(str, prefix) {
    if (str.startsWith(prefix)) {
        return str.slice(prefix.length);
    }
    return str;
}

function getLocales() {
    const url = 'https://capps.capstan.be/langtags_json.php'
    const json = fetch(url, {}).json()
    return Object.values(json).map(entry => entry.BCP47) // object type
}

function searchFileInDirectories(dirPath, folders, filename) {
    // Searches for a file in specified directories
    for (const folder of folders) {
        const directory = path.join(dirPath, folder);
        if (fs.existsSync(directory)) {
            const files = fs.readdirSync(directory);
            if (files.includes(filename) || files.includes(`${filename}${idleExtension}`)) {
                return true;
            }
        }
    }
    return false;
}

function hasNewVersion(filePath, tmxDomain) {
    // Checks if there is a new version of a trend TMX file
    const filename = path.basename(filePath);
    if (filename.includes(trendTag) && filename.includes(tmxDomain)) {
        const newVersionFilename = filename.replace(trendTag, newTag).replace(idleExtension, '');
        const folders = ["tm/auto", "tm/enforce"];
        if (searchFileInDirectories(rootDirPath, folders, newVersionFilename)) {
            return true;
        }
    }
    return false;
}

function getDomain(file) {
    // Extracts the (dirty) domain from a file name; dirty = might contain prefixes/suffixes (CGA-, -New, etc.)
    const fileName = path.basename(file);
    const pattern = /\.tmx(\.zip)?(\.idle)?$/;
    if (fileName.startsWith('PISA_') && pattern.test(fileName)) {
        const tentativeDomain = fileName.split('_')[2];
        if (allowedDomains['QQS'].includes(tentativeDomain) || allowedDomains['QQA'].includes(tentativeDomain)) {
            return Object.keys(allowedDomains).find(key => allowedDomains[key].includes(tentativeDomain)) || null;
        }
        return tentativeDomain;
    } else {
        if (fileName.includes('_QQS_') || fileName.includes('_QQA_')) {
            return fileName.split('_')[1];
        } else if (fileName.includes('_QQSP_') || fileName.includes('_QQAP_')) {
            return removeSuffix(fileName.split('_')[1], "P")
        } else {
            const domain = fileName.split('_')[2];
            return domain.includes('-') ? domain.split('-')[0] : domain;
        }
    }
}

function deleteFile(file) {
    // Deletes a file
    console.log(`>>> Delete ${file.replace(rootDirPath, '')} !!!`);
    try {
        fs.unlinkSync(file);
        console.log(`The file ${file} has been successfully deleted.`);
    } catch (err) {
        if (err.code === 'ENOENT') {
            console.log(`The file ${file} does not exist.`);
        } else {
            console.log(`An error occurred: ${err.message}`);
        }
    }
}

function createDir(dirPath) {
    // Creates a directory
    try {
        fs.mkdirSync(dirPath, { recursive: true });
        console.log(`The folder ${dirPath} and any necessary ancestors in the path have been created.`);
    } catch (err) {
        console.log(`An error occurred: ${err.message}`);
    }
}

function moveFile(origPath, destPath) {
    // Renames file, adding or removing the idle extension
    console.log(`>>> Move ${origPath.replace(rootDirPath, '')} to ${destPath.replace(rootDirPath, '')} !!!`);
    const dirPath = path.dirname(destPath);
    createDir(dirPath);
    try {
        fs.renameSync(origPath, destPath);
        console.log(`The file ${path.basename(destPath)} has been successfully moved.`);
    } catch (err) {
        console.log(`An error occurred: ${err.message}`);
    }
}

function sortRefTmxFileByDomain(filePath, currentDomains) {
    // Sorts reference TMX files by domain
    const dirtyTmxDomain = getDomain(filePath);
    // @todo: refactor this part so that the output of the getDomain() function is already clean (see @strip_domain decorator in the python script)
    const tmxDomain = dirtyTmxDomain.replace("CGA-", "").replace("-New", "").replace("New", "").replace("-Trend", "").replace("Trend", "")

    if (fs.existsSync(filePath)) {
        if (hasNewVersion(filePath, tmxDomain)) {
            if (!filePath.endsWith(idleExtension)) {
                const newFilePath = `${filePath}${idleExtension}`;
                moveFile(filePath, newFilePath);
            }
        } else if (currentDomains.includes(tmxDomain) && filePath.endsWith(idleExtension)) {
            const newFilePath = filePath.replace(idleExtension, '');
            moveFile(filePath, newFilePath);
        } else if (!currentDomains.includes(tmxDomain) && !filePath.endsWith(idleExtension)) {
            const newFilePath = `${filePath}${idleExtension}`;
            moveFile(filePath, newFilePath);
        } else if (disallowedDomains.includes(tmxDomain)) {
            deleteFile(filePath);
        }
    }
}

function getBatchFromFilename(filePath) {
    /* const fileName = path.basename(filePath);
    const fileStem = fileName.split(".")[0];

    if (fileStem.split("_").some(x => locales.includes(x))) {
        // This is a base TM, hence remove language code
        return fileStem.split("_").slice(0, -1).join("_");
    }
    return fileStem; */
    const fileName = path.basename(filePath);
    return batches.filter(batch => fileName.includes(batch))[0];
}

function sortBatchTmxFileByBatch(filePath, batches) {
    // Sorts batch TMX files by batch
    const batch = getBatchFromFilename(filePath)

    if (batches.includes(batch) && filePath.endsWith(idleExtension)) {
        // Remove penalty
        const newFilePath = removeSuffix(filePath, idleExtension);
        moveFile(filePath, newFilePath);
    } else if (!batches.includes(batch) && !filePath.endsWith(idleExtension)) {
        // Add penalty
        const newFilePath = `${filePath}${idleExtension}`;
        moveFile(filePath, newFilePath);
    }
}

function getMappedBatches(rootDirPath) {
    // Retrieves mapped batches
    const settingsFile = path.join(rootDirPath, 'omegat.project');
    const content = fs.readFileSync(settingsFile, 'utf-8');
    const { window } = new JSDOM(content);

    // get the mappings from batch folders in the common repo > source folder
    const mappings = [...window.document.querySelectorAll('mapping[local^="source"]')];
    return mappings.map((mapping) => mapping.getAttribute('local').split('/')[1]);
}

function getBatchDomains(batches) {
    // Retrieves current domains
    return batches.map(getDomain);
}

function getTmxFiles(tmDir, originDirs) {
    // Retrieves TMX files
    const files = originDirs.map(originDir =>
        glob.sync(`${tmDir}/**/${originDir}/*.tmx*`, { recursive: true })
    );
    return files.flat();
}


// Main function
function arrangeTmxFiles(tmDirPath) {
    // Get mapped batches
    const batches = getMappedBatches(rootDirPath);
    // Get current domains
    const currentDomains = getBatchDomains(batches);

    // Trend TMs from previous cycle ('trend') and new TMs from current cycle (called 'ref')
    const refFiles = getTmxFiles(tmDirPath, ["trend", "ref"]);
    refFiles.forEach(tmxFile => {
        sortRefTmxFileByDomain(tmxFile, currentDomains);
    });

    // Batch TMs from previous/next steps + base TMs (also organized by batch) from other locales
    const batchFiles = getTmxFiles(tmDirPath, ["prev", "next", "base", "x-base"]);
    batchFiles.forEach(tmxFile => {
        sortBatchTmxFileByBatch(tmxFile, batches);
    });
}

// Run the main function
arrangeTmxFiles(tmDirPath);

File: code/arrange_tmx_files_with_extension.py
#!/usr/bin/env python3

import os
from pathlib import Path
import sys
import lxml
from bs4 import BeautifulSoup
import re
from glob import glob
from itertools import chain
import fnmatch
import requests
import argparse
from rich import print
from typing import List, Dict, Union, Optional

# type aliases
Response = requests.models.Response
LanguageTags = List[str]

# ############# PROGRAM DESCRIPTION ###########################################

app_desc = "Arrange TMs by batch or domain on batch transition"

# initialize arg parser with a description
parser = argparse.ArgumentParser(description=app_desc)
parser.add_argument("-V", "--version", help="show program version", action="store_true")
parser.add_argument(
    "-r", "--repo", help="specify path to the repository's root directory")

args = parser.parse_args()

# check for -V or --version
version_text = f"{app_desc} v. 0.2"
if args.version:
    print(version_text)
    sys.exit()

if args.repo:
    root_dir_path: str = args.repo.strip()
else:
    print("Required argument not found. Run this script with `--help` for details.")
    sys.exit()

args = parser.parse_args()

# ############# BUSINES LOGIC ###########################################

def get_locales() -> LanguageTags:
    """
    Fetches language tags from an external source.

    Returns:
        list: A list of BCP-47 locales (strings).
    """
    url: str = 'https://capps.capstan.be/langtags_json.php'
    response: Response = requests.get(url)
    return [entry["BCP47"] for entry in response.json()] 


def search_file_in_directories(dir_path: str, folders: list, filename: str) -> bool:
    """
    Searches for a file in specified directories.

    Args:
        dir_path (str): Path to the directory to search in.
        folders (list): List of folders to make the search more specific within the directory.
        filename (str): Name of the file to search for.

    Returns:
        bool: True if the file is found, False otherwise.
    """
    for folder in folders:
        directory: str = os.path.join(dir_path, folder)
        # Iterate through all files and subdirectories in the given directory
        for root, dirs, files in os.walk(directory):
            # Check if the filename exists in the current directory
            if filename in files or f"{filename}{idle_extension}" in files:
                return True
    return False


def has_new_version(file_path: str, tmx_domain: str) -> bool:
    """
    Checks if there is a new version of a trend TMX file, where:
    - trend = produced in the previous cycle
    - new = produced in the current cycle (based on the trend above)

    Args:
        file_path (str): Path to the file.
        tmx_domain (str): Domain associated with the file.

    Returns:
        bool: True if a new version exists, False otherwise.
    """
    filename = os.path.basename(file_path)

    if trend_tag in filename and tmx_domain in filename:
        new_version_fname = filename.replace(trend_tag, new_tag).removesuffix(idle_extension)
        folders = ["tm/auto", "tm/enforce"]
        if search_file_in_directories(root_dir_path, folders, new_version_fname):
            return True
    return False


def strip_domain(affixes):
    def decorator(func):
        def wrapper(*args, **kwargs):
            # Call the original function
            result = func(*args, **kwargs)
            # Check if the result is a string
            if isinstance(result, str):
                # Remove suffixes
                for suffix in affixes['suffix']:
                    if result.endswith(f'-{suffix}'):
                        result = result[: -len(f'-{suffix}')]
                        break
                    elif result.endswith(suffix):
                        result = result[: -len(suffix)]
                        break
                # Remove prefixes
                for prefix in affixes['prefix']:
                    if result.startswith(f'{prefix}-'):
                        result = result[len(f'{prefix}-'):]
                        break
                    elif result.startswith(prefix):
                        result = result[len(prefix):]
                        break
                return result
            else:
                return result
        return wrapper
    return decorator
    

@strip_domain({'suffix': ['New', 'Trend'], 'prefix': ['CGA']})
def get_domain(file: str) -> str:
    """
    Extracts the (dirty) domain from a file name.
    If the domain is dirty, it will be cleaned up by the @strip_domain decorator.

    Args:
        file (str): Name of the file.

    Returns:
        str: Domain extracted from the file name.
    """
    # get file's basename if it's a path (linux specific)
    file_name: str = os.path.basename(file)
    # defines the pattern to match any TMX's extension (which can be zipped or disabled, or not)
    pattern: str = "\.tmx(\.zip)?(\.idle)?$"

    # any files starting with PISA_ contain domain-specific translations
    if file_name.startswith("PISA_") and re.search(pattern, file_name):
        # for tmx files
        tentative_domain: str = file_name.split("_")[2]
        if tentative_domain in allowed_domains["QQS"] or tentative_domain in allowed_domains["QQA"]:
            # then it's not the domain, but the actual questionnaire
            return next(key for key, value in allowed_domains.items() if tentative_domain in value)
        return tentative_domain
    # any other files contain FT2025 batch-specific translations
    else:
        # for batch folders or batch TMs
        if "_QQS_" in file_name or "_QQA_" in file_name: # CBA QQ
            return file_name.split("_")[1]
        elif "_QQSP_" in file_name or "_QQAP_" in file_name: # PBA QQ
            return file_name.split("_")[1].removesuffix("P")
        else:
            return file_name.split("_")[2].split("-")[0] # anything else


def delete_file(file: str) -> None:
    """
    Deletes a file.

    Args:
        file (str): Path to the file.
    """
    print(f">>> Delete {file.replace(root_dir_path, '')} !!!")
    # file = Path(file) if isinstance(file, str) else file # only needed with unlink()
    try:
        # file.unlink()
        os.remove(file)
        print(f"The file {file} has been successfully deleted.")
    except FileNotFoundError:
        print(f"The file {file} does not exist.")
    except Exception as e:
        print(f"An error occurred: {e}")


def create_dir(dir_path: str) -> None:
    """
    Creates a directory.

    Args:
        dir_path (str): Path to the directory.
    """
    try:
        Path(dir_path).mkdir(parents=True, exist_ok=True)
        # print(f"The folder {dir_path} and any necessary ancestors in the path have been created.")
    except Exception as e:
        print(f"An error occurred: {e}")


def move_file(orig_path: str, dest_path: str) -> None:
    """
    Renames file, adding or removing the idle extension.

    Args:
        orig_path (str): Original path of the file.
        dest_path (str): Destination path.
    """
    print(f">>> Move {orig_path.replace(root_dir_path, '')} to {dest_path.replace(root_dir_path, '')} !!!")
    dir_path, file = os.path.split(dest_path)
    create_dir(dir_path)
    try:
        os.replace(orig_path, dest_path)
        print(f"The file {file} has been successfully moved.")
    except FileNotFoundError:
        print(f"The file {file} does not exist.")
    except Exception as e:
        print(f"An error occurred: {e}")


def sort_ref_tmx_file_by_domain(file_path: str, current_domains: list) -> None:
    """
    Sorts reference TMX files by domain. Reference TMs includes trend and new TMs.

    Args:
        file_path (str): Path to the TMX file.
        current_domains (list): List of current domains.
    """
    # gets the domain of a tmx file, according to different file naming patterns
    tmx_domain = get_domain(file_path)

    if os.path.exists(file_path):

        if has_new_version(file_path, tmx_domain):
            # disable any trend TM if it's enabled and that has a new version from the current cycle
            if not file_path.endswith(idle_extension):
                new_file_path = file_path + idle_extension
                move_file(file_path, new_file_path)
        elif tmx_domain in current_domains and file_path.endswith(idle_extension):
            # enables the TM if it's disabled but matches any of the current domains in the  project
            new_file_path = file_path.removesuffix(idle_extension)
            move_file(file_path, new_file_path)
        elif tmx_domain not in current_domains and not file_path.endswith(idle_extension):
            # disables the TM if it's enabled but does not match any of the current domains in the  project
            new_file_path = file_path + idle_extension
            move_file(file_path, new_file_path)
        elif tmx_domain in disallowed_domains:
            # remove the file for good
            delete_file(file_path)


def get_batch_from_filename(file_path: str) -> str:
    """
    Extracts the batch from a file name.

    Args:
        file_path (str): Path to the file.

    Returns:
        str: Batch extracted from the file name.
    """
    file_name: str = os.path.basename(file_path)
    file_stem: str = file_name.split(".")[0]

    if any(x in file_stem.split("_") for x in locales):
        # this is a base TM, hence remove language code
        return "_".join(file_stem.split("_")[0:-1])
    return file_stem


def sort_batch_tmx_file_by_batch(file_path: str, batches: list) -> None:
    """
    Sorts batch TMX files by batch.

    Input can be:

    - Batch TMs, which may come from the previous or the next step.
    - Base TMs (used for borrowing), which come from an advanced step in another workflow.

    The basename of batch TMs (after removing the extension) is already the batch name.
    The basename of base TMs is a concatenation of the batch name and the locale, so both the
    local and the extension need to be removed.

    Args:
        file_path (str): Path to the TMX file.
        batches (list): List of batches.
    """
    batch: str = get_batch_from_filename(file_path)

    if batch in batches and file_path.endswith(idle_extension):
        # remove penalty
        new_file_path: str = file_path.removesuffix(idle_extension)
        move_file(file_path, new_file_path)
    elif batch not in batches and not file_path.endswith(idle_extension):
        # add penalty
        new_file_path: str = file_path + idle_extension
        move_file(file_path, new_file_path)


def get_tmx_files(tm_dir: str, origin_dirs: list) -> list:
    """
    Gets TMX files from specified directories.

    Args:
        tm_dir (str): Path to the TM directory.
        origin_dirs (list): List of directories to search in.

    Returns:
        list: List of paths to TMX files.
    """
    # finds files inside the TM dir having any of the specified origin folders in the file path
    # 'origin' here means where the TM came from, e.g. previous step, previous cycle, another locale, etc.
    files: list = [glob(f"{tm_dir}/**/{origin_dir}/*.tmx*", recursive=True) for origin_dir in origin_dirs]
    # it flattens the list of files before returning it
    return list(chain.from_iterable(files))


def get_batch_domains(batches: list) -> list:
    """
    Gets domains associated with batches (one domain per batch).

    Args:
        batches (list): List of batches.

    Returns:
        list: List of domains associated with the batches.
    """
    # return [get_domain(batch) for batch in batches]
    return list(map(get_domain, batches))


def get_mapped_batches(root_dir_path: str) -> list:
    """
    Gets mapped batches from repository mappings in the omegat project settings.

    Args:
        root_dir_path (str): Path to the root directory of the project.

    Returns:
        list: List of mapped batches.
    """
    # extracts the whole content of the project settings file
    settings_file: str = os.path.join(root_dir_path, "omegat.project")
    with open(settings_file, "r") as f:
        content: str = f.read()
    bs_content: BeautifulSoup = BeautifulSoup(content, "xml")

    # extracts the mapping elements where the value of the 'local' attribute start with 'source'
    mappings: BeautifulSoup.element_classes = bs_content.find_all("mapping", attrs={'local': re.compile(r"^source")})
    # for every mapping, returns the part of the value after 'source/', i.e. the batch name
    return [mapping.attrs['local'].split("/")[1] for mapping in mappings]


if __name__ == "__main__":

    # domains that should not be used, if accidentally added
    disallowed_domains: list = ["CRT", "FLQ", "FNL", "WBQ"]
    # needed domains
    allowed_domains: dict = {
        "QQS": ("STQ", "STQ-UH", "STQ-UO", "ICQ"),
        "QQA": ("SCQ", "TCQ", "PAQ"),
        "COS": ("MAT", "REA", "SCI")
    }

    # these tags identify whether the TM is from current or previous cycle
    trend_tag: str = "MS2022"
    new_tag: str   = "FT2025"

    # path to /tm folder in the repo/project
    tm_dir_path: str = os.path.join(root_dir_path, "tm")

    # batches that are mapped in the project settings
    batches: list = get_mapped_batches(root_dir_path)
    # domains of the currently mapped batches
    current_domains: list = get_batch_domains(batches)
    # extension we add to TMX files to disable them
    idle_extension: str = ".idle"
    # list of language tags
    locales: LanguageTags = get_locales()

    # trend TMs from previous cycle ('trend') and new TMs from current cycle (called 'ref')
    for tmx_file in get_tmx_files(tm_dir_path, ["trend", "ref"]):
        sort_ref_tmx_file_by_domain(tmx_file, current_domains)

    # batch TMs from previous/next steps + base TMs (also organized by batch) from other locales
    for tmx_file in get_tmx_files(tm_dir_path, ["prev", "next", "base", "x-base"]):
        sort_batch_tmx_file_by_batch(tmx_file, batches)
